service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,199 @@
-	if options.DatacenterWatchInterval == 0 {
-		return nil, errors.New("interval cannot be 0")
-	}
 
 	if logger == nil {
 		logger = defaultLogger
 	}
 
-	return &datacenterWatcher{
-		watchInterval: options.DatacenterWatchInterval,
-		logger:        logger,
-		shutdown:      make(chan struct{}),
-		environment:   environment,
-		options:       options,
-	}, nil
+	if options.DatacenterWatchInterval <= 0 {
+		//default consul interval is 5m
+		options.DatacenterWatchInterval = time.Duration(5 * time.Minute)
+	}
 
-}
+	datacenterWatcher := &datacenterWatcher{
+		consulWatchInterval: options.DatacenterWatchInterval,
+		logger:              logger,
+		options:             options,
+		environment:         environment,
+		inactiveDatacenters: make(map[string]bool),
+	}
 
-func (i *datacenterWatcher) start() {
-	go i.watchDatacenters()
-}
+	if options.ChrysomConfig != nil && options.ChrysomConfig.PullInterval > 0 {
 
-func (i *datacenterWatcher) stop() {
-	close(i.shutdown)
-}
+		if environment.Provider() == nil {
+			return nil, errors.New("must pass in a metrics provider")
+		}
 
-func (i *datacenterWatcher) watchDatacenters() {
+		options.ChrysomConfig.MetricsProvider = environment.Provider()
+
+		var datacenterListenerFunc chrysom.ListenerFunc = func(items []model.Item) {
+			updateInactiveDatacenters(items, datacenterWatcher.inactiveDatacenters, &datacenterWatcher.lock, logger)
+		}
 
-	environment := i.environment
-	client := i.environment.Client()
-	logger := i.logger
-	options := i.options
+		options.ChrysomConfig.Listener = datacenterListenerFunc
+
+		options.ChrysomConfig.Logger = logger
+		chrysomClient, err := chrysom.CreateClient(*options.ChrysomConfig)
+
+		if err != nil {
+			return nil, err
+		}
 
-	checkDatacenters := time.NewTicker(i.watchInterval)
+		//create chrysom client and start it
+		datacenterWatcher.chrysomClient = chrysomClient
+		datacenterWatcher.chrysomClient.Start(context.Background())
 
+	} else if options.ChrysomConfig != nil && options.ChrysomConfig.PullInterval <= 0 {
+		return nil, errors.New("chrysom pull interval cannot be 0")
+	}
+
+	//start consul watch
+	ticker := time.NewTicker(datacenterWatcher.consulWatchInterval)
+	go datacenterWatcher.watchDatacenters(ticker)
+
+	return datacenterWatcher, nil
+
+}
+
+func (d *datacenterWatcher) stop() {
+	if d.chrysomClient != nil {
+		d.chrysomClient.Stop(context.Background())
+	}
+}
+
+func (d *datacenterWatcher) watchDatacenters(ticker *time.Ticker) {
 	for {
 		select {
-		case <-i.shutdown:
+		case <-d.environment.Closed():
+			ticker.Stop()
+			d.stop()
 			return
-		case <-checkDatacenters.C:
-			currentInstancers := environment.Instancers()
-			keys := make(map[string]bool)
-			instancersToAdd := make(service.Instancers)
-
-			datacenters, err := getDatacenters(logger, client, options)
+		case <-ticker.C:
+			datacenters, err := getDatacenters(d.logger, d.environment.Client(), d.options)
 
 			if err != nil {
```

Comment:
What happens if we always get an error? Is there any way of knowing that datacenters aren't being updated?

Reply:
update: I would suggest adding a comment here explaining that `getDatacenters()` logs errors itself.  I do think we should add a metric here that tells us either 1) that we got datacenters or 2) the last time we successfully got them.  Otherwise, we won't have any visibility that this isn't working until something else goes wrong.  Adding this metric could be a different pull request.

Reply:
We had a weird issue in caduceus because of some code very similar to this.